### PR TITLE
[Feature] move cold data to object storage without losing any feature(BE)

### PR DESCRIPTION
**Pull request by platoneko**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

This PR supports rowset level data upload on the BE side, so that there can be both cold data and hot data in a tablet, and there is no necessary to prohibit loading new data to cooled tablets.

Each rowset is bound to a `FileSystem`, so that the storage layer can read and write rowsets without perceiving the underlying filesystem.

The abstracted `RemoteFileSystem` can try local caching strategies with different granularity, instead of caching segment files as before.

To avoid conflicts with the code in be/src/io, we temporarily put the file system related code in the be/src/io/fs directory. In the future, `FileReader`s and `FileWriter`s should be unified.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/agent/task_worker_pool.cpp [hunk 1]
```diff
@@ -1683,9 +1701,64 @@ void TaskWorkerPool::_submit_table_compaction_worker_thread_callback() {
     }
 }
 
-void TaskWorkerPool::_storage_medium_migrate_v2_worker_thread_callback() {
+void TaskWorkerPool::_storage_refresh_storage_policy_worker_thread_callback() {
```
Comment (pengxiangyu) [inline be/src/agent/task_worker_pool.cpp:1704]: why do you delete _storage_medium_migrate_v2_worker_thread_callback？do you want to implement it in BE?

File: be/src/olap/tablet.cpp [hunk 2]
```diff
@@ -1636,4 +1652,160 @@ std::shared_ptr<MemTracker>& Tablet::get_compaction_mem_tracker(CompactionType c
...
+    if (!cumu_compaction_lock.owns_lock()) {
+        LOG(WARNING) << "cumulative compaction is running. tablet=" << tablet_id();
+        return Status::OLAPInternalError(OLAP_ERR_BE_TRY_BE_LOCK_ERROR);
+    }
+    auto dest_fs = io::FileSystemMap::instance()->get(cooldown_resource());
+    if (!dest_fs) {
+        return Status::OLAPInternalError(OLAP_ERR_NOT_INITED);
+    }
+    DCHECK(dest_fs->type() == io::FileSystemType::S3);
+    auto old_rowset = pick_cooldown_rowset();
```
Comment (pengxiangyu) [inline be/src/olap/tablet.cpp:1677]: why only one rowset? how abort the others?

File: be/src/olap/tablet.cpp [hunk 3]
```diff
@@ -1636,4 +1652,160 @@ std::shared_ptr<MemTracker>& Tablet::get_compaction_mem_tracker(CompactionType c
...
+                rowset = rs;
+            }
+        }
+    }
+    return rowset;
+}
+
+bool Tablet::need_cooldown(int64_t* cooldown_timestamp, size_t* file_size) {
+    // std::shared_lock meta_rlock(_meta_lock);
+    if (cooldown_resource().empty()) {
```
Comment (pengxiangyu) [inline be/src/olap/tablet.cpp:1733]: cooldown_resource()?  isn't it policy_name?

File: be/src/olap/tablet_manager.cpp [hunk 4]
```diff
@@ -1345,18 +1271,55 @@ void TabletManager::get_tablets_distribution_on_different_disks(
...
 
+struct SortCtx {
+    SortCtx(TabletSharedPtr tablet, int64_t cooldown_timestamp, int64_t file_size)
+            : tablet(tablet), cooldown_timestamp(cooldown_timestamp), file_size(file_size) {}
+    TabletSharedPtr tablet;
+    int64_t cooldown_timestamp;
+    int64_t file_size;
+};
+
+void TabletManager::get_cooldwon_tablets(std::vector<TabletSharedPtr>* tablets) {
```
Comment (pengxiangyu) [inline be/src/olap/tablet_manager.cpp:1287]: cooldown, misspelling

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -304,29 +270,28 @@ RowsetSharedPtr BetaRowsetWriter::build() {
...
-    // and tablets with the same type should share one BlockManager object;
-    fs::BlockManager* block_mgr = fs::fs_util::block_manager(_context.path_desc);
-    std::unique_ptr<fs::WritableBlock> wblock;
-    fs::CreateBlockOptions opts(path_desc);
-    DCHECK(block_mgr != nullptr);
-    Status st = block_mgr->create_block(opts, &wblock);
+    auto path = BetaRowset::local_segment_path(_context.tablet_path, _context.rowset_id,
+                                               _num_segment++);
+    auto fs = _rowset_meta->fs();
+    if (!fs) {
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/olap/rowset/beta_rowset_writer.cpp:276]: We print WARNING LOG in RowsetMeta->fs()

File: be/src/olap/base_compaction.cpp [hunk 6]
```diff
@@ -106,14 +106,11 @@ Status BaseCompaction::pick_rowsets_to_compact() {
...
+    // `_input_rowsets` has been sorted by end version, so we consider `_input_rowsets[0]` is the base rowset.
+    int64_t base_size = _input_rowsets.front()->data_disk_size();
     int64_t cumulative_total_size = 0;
-    for (auto& rowset : _input_rowsets) {
-        if (rowset->start_version() != 0) {
-            cumulative_total_size += rowset->data_disk_size();
-        } else {
-            base_size = rowset->data_disk_size();
-        }
+    for (auto it = _input_rowsets.begin() + 1; it != _input_rowsets.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/olap/base_compaction.cpp:112]: _input_rowsets has been sorted at line 89, and _check_rowset_overlapping at line 91. So i think it has been in order.

File: be/src/olap/tablet.cpp (hunk 3 above)
Comment (platoneko) [reply be/src/olap/tablet.cpp:1733]: yeah, i will change it to storage_policy soon,

**Comment by dataroaring** [comment]
[冷热分离详细设计.pdf](https://github.com/apache/doris/files/9033096/default.pdf) @pengxiangyu

File: fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java [hunk 7]
```diff
@@ -213,7 +222,7 @@ public static DataProperty analyzeDataProperty(Map<String, String> properties, D
         }
 
         Preconditions.checkNotNull(storageMedium);
-        return new DataProperty(storageMedium, cooldownTimeStamp, remoteStorageResourceName, remoteCooldownTimeStamp);
+        return new DataProperty(storageMedium, cooldownTimeStamp, remoteStorageResourceName, remoteCooldownTimeStamp, storagePolicy);
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/common/util/PropertyAnalyzer.java:225]: If MigrationHandler is on BE, remoteStorageResourceName and remoteCooldownTimeStamp are not useful, just remove them from DataProperty.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/TableProperty.java [hunk 8]
```diff
@@ -139,6 +141,15 @@ public TableProperty buildInMemory() {
         return this;
     }
 
+    public TableProperty buildStoragePolicy() {
+        storagePolicy = properties.getOrDefault(PropertyAnalyzer.PROPERTIES_STORAGE_POLICY, "");
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/TableProperty.java:145]: default is ""？How abort use default_storage_policy in Config？

File: fe/fe-core/src/main/java/org/apache/doris/catalog/StoragePolicyResource.java [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+ * Syntax:
+ * CREATE RESOURCE "storage_policy_name"
+ * PROPERTIES(
+ *      "type"="storage_policy",
+ *      "cooldown_datetime" = "2022-06-01", // time when data is transfter to medium
+ *      "cooldown_ttl" = "3600"， // data is transfter to medium after 1 hour
+ *      "s3_*"
+ * );
+ */
+public class StoragePolicyResource extends Resource {
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/StoragePolicyResource.java:54]: This class is replaced by StoragePolicy, as discuss before, it is a policy, but not a resource.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/ResourceMgr.java [hunk 10]
```diff
@@ -82,6 +85,20 @@ public void createResource(CreateResourceStmt stmt) throws DdlException {
         LOG.info("Create resource success. Resource: {}", resource);
     }
 
+    public void createDefaultStoragePolicy() throws DdlException {
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/ResourceMgr.java:88]: StoragePolicyResource is not needed. Use StoragePolicy instead.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/DataProperty.java [hunk 11]
```diff
@@ -46,6 +46,8 @@ public class DataProperty implements Writable {
     private String remoteStorageResourceName;
```
Comment (pengxiangyu) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/DataProperty.java:46]: If Migration Handler is in BE, remoteStorageResourceName and remoteCooldownTimeMs are not needed, just remove them. The serialization for this class is a json.

File: fe/fe-core/src/main/java/org/apache/doris/task/DropReplicaTask.java [hunk 12]
```diff
@@ -23,11 +23,13 @@
 public class DropReplicaTask extends AgentTask {
     private int schemaHash; // set -1L as unknown
     private long replicaId;
+    private boolean isDropTableOrPartition;
```
[the remark this answers is not in the document]
Comment (platoneko) [reply fe/fe-core/src/main/java/org/apache/doris/task/DropReplicaTask.java]: FE should distinguish this DropReplicaTask is to "drop table/partition" or just "drop replica". When "drop replica", BE should not delete rowsets on remote storage, as these rowsets may be used by other BEs.

File: be/src/agent/task_worker_pool.cpp [hunk 13]
```diff
@@ -437,6 +443,11 @@ void TaskWorkerPool::_drop_tablet_worker_thread_callback() {
             StorageEngine::instance()->txn_manager()->force_rollback_tablet_related_txns(
                     dropped_tablet->data_dir()->get_meta(), drop_tablet_req.tablet_id,
                     drop_tablet_req.schema_hash, dropped_tablet->tablet_uid());
+            // We remove remote rowset directly.
+            // TODO(cyx): do remove in background
+            if (drop_tablet_req.is_drop_table_or_partition) {
+                dropped_tablet->remove_all_remote_rowsets();
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/agent/task_worker_pool.cpp:448]: We will solve this in next PR.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**